neuralmonkey/decoders/transformer.py

```diff
@@ -0,0 +1,347 @@
+                queries=queries,
+                keys=encoder_att_states,
+                values=encoder_att_states,
+                keys_mask=encoder_att_mask,
+                num_heads=self.n_heads_enc,
+                dropout_callback=lambda x: dropout(
+                    x, self.attention_dropout_keep_prob, self.train_mode))
+
+            return dropout(
+                encoder_context, self.dropout_keep_prob, self.train_mode)
+
+    def layer(self, level: int, inputs: tf.Tensor,
+              mask: tf.Tensor) -> TransformerLayer:
+        # Recursive implementation. Outputs of the zeroth layer are the inputs
+        if level == 0:
+            norm_inputs = tf.contrib.layers.layer_norm(
+                inputs, begin_norm_axis=2)
+            return TransformerLayer(norm_inputs, mask)
+
+        # Compute the outputs of the previous layer
+        prev_layer = self.layer(level - 1, inputs, mask)
+
+        # Run self-attention
+        self_context = self.masked_self_attention(level, prev_layer)
+
+        # Residual connections + layer normalization
+        encoder_queries = tf.contrib.layers.layer_norm(
+            self_context + prev_layer.temporal_states, begin_norm_axis=2)
+
+        # Attend to the encoder
+        encoder_context = self.encoder_attention(level, encoder_queries)
+
+        # Residual connections + layer normalization
+        ff_input = tf.contrib.layers.layer_norm(
+            encoder_context + self_context, begin_norm_axis=2)
+
+        # Feed-forward network hidden layer + ReLU + dropout
+        ff_hidden = tf.layers.dense(
+            ff_input, self.ff_hidden_size, activation=tf.nn.relu,
+            name="ff_hidden_{}".format(level))
+        ff_hidden_drop = dropout(
+            ff_hidden, self.dropout_keep_prob, self.train_mode)
+
+        # Feed-forward output projection + dropout
+        ff_output = tf.layers.dense(
+            ff_hidden_drop, self.dimension, name="ff_out_{}".format(level))
+        ff_output = dropout(ff_output, self.dropout_keep_prob, self.train_mode)
+
+        # Residual connections + layer normalization
+        output_states = tf.contrib.layers.layer_norm(
+            ff_output + ff_input, begin_norm_axis=2)
+
+        return TransformerLayer(states=output_states, mask=mask)
+
+    @tensor
+    def train_logits(self) -> tf.Tensor:
+        last_layer = self.layer(self.depth, self.embedded_train_inputs,
+                                tf.transpose(self.train_mask))
+        # matmul with output matrix
+        # t_states shape: (batch, time, channels)
+        # dec_w shape: (channels, vocab)
+        logits_unbiased = tf.nn.conv1d(
+            last_layer.temporal_states, tf.expand_dims(self.decoding_w, 0), 1,
+            "SAME")
+
+        # shape (batch, time, vocab)
+        logits = logits_unbiased + tf.reshape(self.decoding_b, [1, 1, -1])
+
+        # return logits in time-major shape
+        return tf.transpose(logits, perm=[1, 0, 2])
+
+    # @tensor
```

Comment:
Co je tohle? Vskutečnosti v předkovi?

Reply:
alternativní implementace `train_xents` z předka s label smoothing

Reply:
dark slide

Reply:
takze smazat?

Reply:
Dělalo to někdy něco? Když už je ten kód napsaný, tak bych přidal někam (možná do předka?) nějaký option na potenciální label smoothing.

Reply:
dal bych to do předka a přidal tomu option a na default to hodil vypnutý. Je ta úprava triviální? Jestli ne, tak bych to mergnul bez toho, potažmo s timhle zakomentovaným - v T2T ten label smoothing je, takže budeme-li chtít to replikovat, budeme potřebovat i tuhle featuru

Reply:
ok, tenhle "dark" label smoothing hodim do neuralmonkey.functions (otazka je, jaky tomu dat kryci nazev), do predka pridam option label_smoothing_function, jako default bude mit default z tf.contrib.seq2seq.sequence_loss